connectors/sources/salesforce.py

```diff
@@ -183,23 +203,35 @@ async def close(self):
             del self.session
 
     async def get_docs(self):
-        async for account in self.get_accounts():
-            yield account, None
+        all_content_docs = []
 
-        async for opportunity in self.get_opportunities():
-            yield opportunity, None
+        async for account, content_docs in self.get_accounts():
+            all_content_docs.extend(content_docs)
+            yield account
 
-        async for contact in self.get_contacts():
-            yield contact, None
+        async for opportunity, content_docs in self.get_opportunities():
+            all_content_docs.extend(content_docs)
+            yield opportunity
 
-        async for lead in self.get_leads():
-            yield lead, None
+        async for contact, content_docs in self.get_contacts():
+            all_content_docs.extend(content_docs)
+            yield contact
 
-        async for campaign in self.get_campaigns():
-            yield campaign, None
+        async for lead, content_docs in self.get_leads():
+            all_content_docs.extend(content_docs)
+            yield lead
 
-        async for case in self.get_cases():
-            yield case, None
+        async for campaign, content_docs in self.get_campaigns():
+            all_content_docs.extend(content_docs)
+            yield campaign
+
+        async for case, content_docs in self.get_cases():
+            all_content_docs.extend(content_docs)
+            yield case
+
+        all_content_docs = self._combine_duplicate_content_docs(all_content_docs)
```

Comment:
How big are the objects?

How much is the memory growth when all of that is stored in-memory?

If it's big enough, then we probably need to do it on the fly.

Reply:
I guess we have no idea, a user could conceivably have a massive amount of files.

> we probably need to do it on the fly.

Do you mean check for duplicates as we're downloading files?

Reply:
> I guess we have no idea, a user could conceivably have a massive amount of files.

That's something good to test, when we have a proper test instance.

> Do you mean check for duplicates as we're downloading files?

I mean that right now you do the following: collect ids for objects A, B, C, D...Z, then combine and de-dupe. To optimise it (if duplicates are actually happening often, of course) you can do it on the fly, imagine the following class:

```python
class IdDeduplicator:
    def __init__(self):
        self.ids = {}

    def add(obj):
        if self._has_id(obj):
            self._append_to_existing(obj)
        else:
            self._append_non_existing(obj)

    def get():
        return self.ids
```

So every time an object is added, deduplication happens. We'll trade memory vs CPU here, but it's efficient as long as there are a lot of duplicate objects. I believe there's no need to touch it now, but good to keep this in mind for later testing. Also, something to collect for telemetry if we're not sure about actual numbers of duplicates even after testing.

Reply:
Ah I see! Thanks for the summary. I'm going to open a few improvement issues for this connector so I'll include both of your points here (optimising this and ensuring we get telemetry on it).